Make unbind() overrideable for NT subclass

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #109123
* __->__ #109122

Goal: avoid making unbind composite implicit so we can override it within `__torch_dispatch__()` for the NT subclass.

jbschlosser committed:
Make unbind() overrideable for NT subclass

[ghstack-poisoned]

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -2786,8 +2786,13 @@
+      self: unbind_backward(grads, dim)
+      result: auto_linear
+    AutogradNestedTensor:
+      self: at::stack(grads, dim)
```
If that worked, we would use it :p
I think that proper use of `zeros_like()` in unbind_backward should do the trick?

jbschlosser replied:
@albanD I added a separate `unbind_backward_nested` to cover this since we need to store the full NT shape info to "stack" properly. My approach stores `result` (a `vector<Tensor>`) for backward, and interestingly this hit some codegen asserts assuming foreach ops when storing a vector result. If I tweak those asserts, things seem to still work. Thoughts on the right way to handle this?

jbschlosser committed:
Update on "Make unbind() overrideable for NT subclass"


Goal: avoid making unbind composite implicit so we can override it within `__torch_dispatch__()` for the NT subclass.

[ghstack-poisoned]

jbschlosser committed:
Update on "Make unbind() overrideable for NT subclass"


Goal: avoid making unbind composite implicit so we can override it within `__torch_dispatch__()` for the NT subclass.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

huydhn commented:
@jbschlosser I think this change is causing a memory leak https://hud.pytorch.org/pytorch/pytorch/commit/c71a64ccce7215ab9b3938582c6249b62a72be8b.  I could reproduce it locally with:

```
PYTORCH_TEST_CUDA_MEM_LEAK_CHECK=1 PYTORCH_TEST_WITH_SLOW_GRADCHECK=1 python test/functorch/test_vmap.py -k test_fallback_binary_cuda
```

or 

```
PYTORCH_TEST_CUDA_MEM_LEAK_CHECK=1 PYTORCH_TEST_WITH_SLOW_GRADCHECK=1 python test/test_nestedtensor.py -k test_backward_sub_strided_cuda
```

Reverting this change locally and the leak issue is fixed.  Could you help take a look?

PaliC commented:
@pytorchbot revert -m "breaking slow tests" -c "nosignal"

https://hud.pytorch.org/pytorch/pytorch/commit/7f5737392d637a22d555a88a8546d8fc7ab31084 is an example of the broken slow tests. Unfortunately, the hud page for the commit doesn't show them as another PR made the tests even more broken :(
